plugins/jobtype/src/azkaban/jobtype/HadoopSecureSparkWrapper.java

```diff
@@ -194,39 +206,115 @@ private static void handleDynamicResourceAllocation(String[] argArray) {
+               argArray[i+1].startsWith(SPARK_CONF_QUEUE))
+              || (argArray[i].equals(SparkJobArg.QUEUE.sparkParamName))) {
+
+            logger.info("Azbakan enforces node labeling. Ignore user param: "
+              + argArray[i] + " " + argArray[i+1]);
+            argArray[i] = null;
+            argArray[++i] = null;
+            continue;
+          }
+          if (autoNodeLabeling) {
+            // If auto node labeling is enabled, job type should ignore user supplied
+            // node label expression for Spark executors. This config will be automatically
+            // set by the job type based on the mem-to-vcore resource ratio requested by
+            // the user application.
+            if (argArray[i].equals(SparkJobArg.SPARK_CONF_PREFIX.sparkParamName) &&
+                argArray[i+1].startsWith(SPARK_EXECUTOR_NODE_LABEL_EXP)) {
+              logger.info("Azbakan auto-sets node label expression. Ignore user param: "
+                + argArray[i] + " " + argArray[i+1]);
+              argArray[i] = null;
+              argArray[++i] = null;
+              continue;
+            }
+            if (argArray[i].equals(SparkJobArg.EXECUTOR_CORES.sparkParamName)) {
+              executorVcore = argArray[++i];
+            }
+            if (argArray[i].equals(SparkJobArg.EXECUTOR_MEMORY.sparkParamName)) {
+              executorMem = argArray[++i];
+            }
+            if (argArray[i].equals(SparkJobArg.SPARK_CONF_PREFIX.sparkParamName) &&
+                argArray[i+1].startsWith(SPARK_EXECUTOR_MEMORY_OVERHEAD)) {
+              executorMemOverhead = argArray[i+1].split("=")[1].trim();
+            }
+          }
         }
       }
+      // If auto node labeling is enabled, automatically sets spark.yarn.executor.nodeLabelExpression
+      // config based on user requested resources.
+      if (autoNodeLabeling) {
+        double minRatio = Double.parseDouble(minMemVcoreRatio);
+        if (executorVcore == null) {
+          executorVcore = sparkConf.get(SPARK_EXECUTOR_CORES, SPARK_EXECUTOR_DEFAULT_CORES);
+        }
+        if (executorMem == null) {
+          executorMem = sparkConf.get(SPARK_EXECUTOR_MEMORY, SPARK_EXECUTOR_DEFAULT_MEMORY);
+        }
+        if (executorMemOverhead == null) {
+          executorMemOverhead = sparkConf.get(SPARK_EXECUTOR_MEMORY_OVERHEAD, null);
+        }
+        if (calculateMemVcoreRatio(executorMem, executorMemOverhead, executorVcore,
+            sparkConf, conf) > minRatio) {
+          LinkedList<String> argList = new LinkedList<String>(Arrays.asList(argArray));
+          argList.addFirst(SPARK_EXECUTOR_NODE_LABEL_EXP + "=" + desiredNodeLabel);
+          argList.addFirst(SparkJobArg.SPARK_CONF_PREFIX.sparkParamName);
+          argArray = argList.toArray(new String[argList.size()]);
+        }
+      }
+    }
+    return argArray;
+  }
+
+  private static double calculateMemVcoreRatio(String mem, String memOverhead, String vcore,
```

Comment:
Is there a test that covers this code path? i.e. verify the calculation is correct?